tests/fuzz/s2n_server_cert_recv_test.c

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+/* Target Functions: s2n_server_cert_recv s2n_x509_validator_validate_cert_chain */
+
+#include <errno.h>
+#include <fcntl.h>
+#include <stddef.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+
+#include <openssl/crypto.h>
+#include <openssl/err.h>
+
+#include "api/s2n.h"
+#include "stuffer/s2n_stuffer.h"
+#include "tls/s2n_cipher_suites.h"
+#include "tls/s2n_config.h"
+#include "tls/s2n_connection.h"
+#include "tls/s2n_crypto.h"
+#include "tls/s2n_tls.h"
+#include "tls/s2n_tls_parameters.h"
+#include "utils/s2n_random.h"
+#include "utils/s2n_safety.h"
+
+#include "s2n_test.h"
+#include "testlib/s2n_testlib.h"
+
+struct host_verify_data {
+    const char *name;
+    uint8_t found_name;
+    uint8_t callback_invoked;
+};
+
+static uint8_t verify_host_accept_everything(const char *host_name, size_t host_name_len, void *data) {
+    struct host_verify_data *verify_data = (struct host_verify_data *) data;
+    verify_data->callback_invoked = 1;
+    return 1;
+}
+
+static const uint8_t TLS_VERSIONS[] = {S2N_TLS10, S2N_TLS11, S2N_TLS12, S2N_TLS13};
+
+static void s2n_fuzz_atexit()
+{
+    s2n_cleanup();
+}
+
+int LLVMFuzzerInitialize(const uint8_t *buf, size_t len)
+{
+#ifdef S2N_TEST_IN_FIPS_MODE
+    S2N_TEST_ENTER_FIPS_MODE();
+#endif
+
+    GUARD(s2n_init());
+    GUARD_STRICT(atexit(s2n_fuzz_atexit));
+    return 0;
+}
+
+int LLVMFuzzerTestOneInput(const uint8_t *buf, size_t len)
+{
+    /* We need at least one byte of input to set parameters */
+    if (len != 0) {
+
+        /* Setup */
+        struct host_verify_data verify_data = { .callback_invoked = 0, .found_name = 0, .name = NULL };
+        struct s2n_x509_trust_store trust_store;
+        struct s2n_connection *conn = s2n_connection_new(S2N_CLIENT);
+        notnull_check(conn);
+        GUARD(s2n_stuffer_write_bytes(&conn->handshake.io, buf, len));
+
+        /* Returns void, so can't be guarded */
+        s2n_x509_validator_wipe(&conn->x509_validator);
+        s2n_x509_trust_store_init_empty(&trust_store);
+
+        /* Pull a byte off the libfuzzer input and use it to set parameters */
+        uint8_t randval = 0;
+        GUARD(s2n_stuffer_read_uint8(&conn->handshake.io, &randval));
```

Comment:
Using 1 byte of LibFuzzer's input will work, but it ends up doubling the number of corpus files stored, and also means that it's possible that LibFuzzer could miss an interesting input if the first byte had the "wrong" value but the remaining bytes would have reached a new branch.

I'd rather use a `for` loop that checks both options instead of letting LibFuzzer decide.

Reply:
After the update this actually has greater code coverage than before! This is also true across the board, I've applied it to all the fuzz testing functions in what's soon to be a different PR. LibFuzzer still explores all possible branches, just in different fuzz tests, and this lets it dig deeper into each instead of trying to optimize coverage across multiple branches at once. The number of corpus files does go up, but that's the intended behavior of LibFuzzer and lets it more effectively cover code.

Reply:
After speaking with Audrey offline, it seems that doing it this way actually increases LibFuzzer code coverage. Our current theory is that performing multiple test runs in a for loop in a single fuzz test leads to noisier coverage measurements for LibFuzzer, which makes it harder for LibFuzzer figure out what byte change resulted in which branch coverage change.